cache/lru_sized_cache.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package cache
+
+import (
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils"
+	"github.com/ava-labs/avalanchego/utils/linkedhashmap"
+)
+
+var _ Cacher[struct{}, SizedElement] = (*sizedLRU[struct{}, SizedElement])(nil)
+
+type SizedElement interface {
+	Size() int
+}
+
+// sizedLRU is a key value store with bounded size. If the size is attempted to
+// be exceeded, then elements are removed from the cache until the bound is
+// honored, based on evicting the least recently used value.
+type sizedLRU[K comparable, V SizedElement] struct {
+	lock        sync.Mutex
+	elements    linkedhashmap.LinkedHashmap[K, V]
+	maxSize     int
+	currentSize int
+}
+
+func NewSizedLRU[K comparable, V SizedElement](maxSize int) Cacher[K, V] {
+	return &sizedLRU[K, V]{
+		elements: linkedhashmap.New[K, V](),
+		maxSize:  maxSize,
+	}
+}
+
+func (c *sizedLRU[K, V]) Put(key K, value V) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.put(key, value)
+}
+
+func (c *sizedLRU[K, V]) Get(key K) (V, bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	return c.get(key)
+}
+
+func (c *sizedLRU[K, V]) Evict(key K) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.evict(key)
+}
+
+func (c *sizedLRU[K, V]) Flush() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.flush()
+}
+
+func (c *sizedLRU[K, V]) put(key K, value V) {
+	if oldValue, ok := c.elements.Get(key); ok {
+		c.currentSize -= oldValue.Size()
+	}
+
+	c.elements.Put(key, value)
+	c.currentSize += value.Size()
```

Comment:
currentSize could overflow maxSize here - should we remove elements first before putting?

Reply:
Yeah, that was my preference too

Reply:
I don't understand. If `value.Size() > c.maxSize` then the cache should be empty at the end - no? Otherwise we aren't actually honoring `maxSize`

Reply:
`value.Size()` could be larger than `c.maxSize - c.currentSize` which would cause it to overflow.

e.g. if `currentSize` is `15`, `maxSize` is `20` and you're putting a new key with `value.Size()` of `6`.

Reply:
@dhrubabasu should be good now